### PR TITLE
Teleportation environment

### PR DESCRIPTION
**Pull request by jwallnoefer**
This pull request adds an environment in which the agent is expected to learn the quantum teleportation protocol.

### TODO
- [x]  Get tests to pass.
- [x]  Write README
- [x]  Wait until testing branch is merged with master.

------------------
Before merging I will need to expand the README to better discuss the purpose and limitations of the environment. But for now I would kindly ask you to review whether I adapted my environment correctly to match your standard.

### CODE REVIEW

**Comment by HendrikPN** [comment]
Thanks a lot for the environment! We are very happy for it. I will have a closer look soon, but at first glance all looks fine. We are about to include unit tests to scigym. If you want, you can run you environment against the tests. You can find the current tests on this [branch](https://github.com/HendrikPN/scigym/tree/issue_3_testing)

UPDATE: Currently, however it seems our tests run into problems with your environment, but that might just be an error on our side. The error is,
```
TypeError: '<=' not supported between instances of 'int' and 'tuple'
```
which seems to be cause by your `MultiDiscrete` observation space.

**Comment by jwallnoefer** [comment]
I managed to get the tests to pass. It was indeed me violating some scigym standards!

I'm not sure if merging this branch with the testing branch was the best way to run the tests on my environment. - Please let me know if there's a better way that doesn't look so confusing.

On second thought... I think with the merge I somehow destroyed the github history of the branches... oh well...

**Comment by HendrikPN** [comment]
Yeh. Maybe wait for the testing branch to be merged to master before merging the master into your branch.

**Comment by jwallnoefer** [comment]
Ok... I'll do that!

**Comment by HendrikPN** [comment]
#6 is resolved, you can include the testing from the master now!

File: setup.py [hunk 1]
```diff
@@ -7,7 +7,8 @@
 
 # Environment-specific dependencies.
 extras = {
-    'surfacecode': ['tensorflow==1.5', 'keras==2.2.4']
+    'surfacecode': ['tensorflow==1.5', 'keras==2.2.4'],
+    'teleportation': ["numpy"]
```
Comment (HendrikPN) [inline setup.py]: You don't need to specify `numpy`. It is assumed to be available.

File: scigym/envs/quantum_physics/quantum_computing/teleportation/README.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+*Observations*
+The observations perceived by the agent are a list of previous actions and measurement outcomes. Invalid actions (on qubits that have already been measured) are not added to this list.
+
+
+#### Known solutions
+
+#### Implementation details
+
+#### Discussion
```
Comment (HendrikPN) [inline scigym/envs/quantum_physics/quantum_computing/teleportation/README.md:76]: I suppose you will improve on this before merging!

**Comment by jwallnoefer** [comment]
The environment is finished from my side. Please let me know if anything needs changing or if anything is unclear from the Documentation.

File: scigym/envs/quantum_physics/quantum_computing/teleportation/README.md [hunk 3]
```diff
@@ -22,17 +22,71 @@ where the input qubit is an arbitrary state Ψ.
...
+In order to make sure that the solution does not depend on the input state,
+we compare the Jamiolkowski fidelity of the effective channel generated by the
+gate sequence picked by the agent with the ideal channel. We do this by
+initializing the qubit A' in a state where it is entangled with an
+auxiliary qubit in a maximally entangled state Φ<sup>+</sup>.
+
+*Careful:* Because the environment will return `done` as `True` once no valid
+actions remain, this can lead to very short, but unrewarded trials.
+
 #### Discussion
```
Comment (HendrikPN) [inline scigym/envs/quantum_physics/quantum_computing/teleportation/README.md:76]: Do you not want to add your paper reference somewhere?

**Review by HendrikPN — APPROVED** [review]

File: scigym/envs/quantum_physics/quantum_computing/teleportation/README.md (hunk 3 above)
Comment (jwallnoefer) [reply scigym/envs/quantum_physics/quantum_computing/teleportation/README.md:76]: That's probably a good idea.

**Review by HendrikPN — APPROVED** [review]

**Pull request merged**